File: python/110_Balanced_Binary_Tree.py
# coding=utf-8
# #problem number 110
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        
class Solution:
    # @param {TreeNode} root
    # @return {boolean}
    def isBalanced(self, root):
        if root is None or (root.left is None and root.right is None):
            return True
        if self.isBalanced(root.left) and self.isBalanced(root.right) and -1 <= (self.height(root.left) - self.height(root.right)) <= 1:
            return True
        return False
        
    def height(self, node):
        if node is None:
            return 0
        return 1+max(self.height(node.left),self.height(node.right))
        
s=Solution()
root1=TreeNode(2)
root1.left=TreeNode(1)
root1.right=TreeNode(4)
root1.right.left=TreeNode(3)
root1.right.right=TreeNode(6)
root1.right.right.left=TreeNode(5)
print((s.isBalanced(root1)))
root2=TreeNode(72)
root2.left=TreeNode(54)
root2.left.right=TreeNode(67)
root2.right=TreeNode(76)
print((s.isBalanced(root2)))

File: Java/110_Balanced_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public boolean isBalanced(TreeNode root) {
        return !(balanced(root)==-1);
    }
    
    // 返回高度的同时，以一个特殊的值返回该子树是否平衡，避免重复计算
    public int balanced(TreeNode root){
        if(root==null) return 0;
        int left=balanced(root.left);
        int right=balanced(root.right);
        if(left==-1||right==-1||Math.abs(left-right)>1){//若某个子树不平衡或者当前不平衡则返回-1
            return -1;
        }else{
            return Math.max(left,right)+1;
        }
    }
}